command/agent/agent_endpoint_test.go

```diff
@@ -672,60 +550,43 @@ func TestAgent_Leave(t *testing.T) {
 }
 
 func TestAgent_Leave_ACLDeny(t *testing.T) {
-	// Try without a token.
-	func() {
+	t.Run("no token", func(t *testing.T) {
 		dir, srv := makeHTTPServerWithACLs(t)
 		defer os.RemoveAll(dir)
 		defer srv.Shutdown()
 		defer srv.agent.Shutdown()
 
-		req, err := http.NewRequest("PUT", "/v1/agent/leave", nil)
-		if err != nil {
-			t.Fatalf("err: %v", err)
-		}
-
-		_, err = srv.AgentLeave(nil, req)
-		if err == nil || !strings.Contains(err.Error(), permissionDenied) {
+		req, _ := http.NewRequest("PUT", "/v1/agent/leave", nil)
+		if _, err := srv.AgentLeave(nil, req); !isPermissionDenied(err) {
 			t.Fatalf("err: %v", err)
 		}
-	}()
+	})
 
-	// Try the agent master token (resolved on the agent).
-	func() {
+	t.Run("agent master token", func(t *testing.T) {
 		dir, srv := makeHTTPServerWithACLs(t)
 		defer os.RemoveAll(dir)
 		defer srv.Shutdown()
 		defer srv.agent.Shutdown()
 
-		req, err := http.NewRequest("PUT", "/v1/agent/leave?token=towel", nil)
-		if err != nil {
-			t.Fatalf("err: %v", err)
-		}
-
-		_, err = srv.AgentLeave(nil, req)
-		if err != nil {
+		req, _ := http.NewRequest("PUT", "/v1/agent/leave?token=towel", nil)
+		if _, err := srv.AgentLeave(nil, req); err != nil {
 			t.Fatalf("err: %v", err)
 		}
-	}()
+	})
 
-	// Try with a read only token (resolved on the servers).
-	func() {
+	// this test fails when all subtests share the same server.
```

Comment:
I couldn't get it to work when moving the test server setup out and started only one server for all tests. It reliably failed. Note that this is how it was before as well and that's the only test written like this.